From 5867635f437097002ea20752e02c9fdea156e8ab Mon Sep 17 00:00:00 2001
From: Henrik Grimler <henrik@grimler.se>
Date: Mon, 8 Jul 2024 21:27:16 +0200
Subject: [PATCH] heimdall: fix segfault when flashing with parition identifier

When checking the size of the image and partition we first need to
ensure the partition exist, and handle the case of passing the
partition identifier instead of partition name.

Fixes: 60ab9bbaffe3 ("FlashAction: Make sure file fit partition before flashing")
---
 heimdall/source/FlashAction.cpp | 24 +++++++++++++++++++++++-
 1 file changed, 23 insertions(+), 1 deletion(-)

diff --git a/heimdall/source/FlashAction.cpp b/heimdall/source/FlashAction.cpp
index 4ea39b0696d6..52790b9b8d0f 100644
--- a/heimdall/source/FlashAction.cpp
+++ b/heimdall/source/FlashAction.cpp
@@ -301,7 +301,29 @@ static bool flashPartitions(BridgeManager *bridgeManager, const vector<Partition
 	{
 		for (vector<PartitionFile>::const_iterator it = partitionFiles.begin(); it != partitionFiles.end(); it++)
 		{
-			const PitEntry *part = pitData->FindEntry(it->argumentName);
+			unsigned int partitionIdentifier;
+			const PitEntry *part;
+			if (Utility::ParseUnsignedInt(partitionIdentifier, it->argumentName) == kNumberParsingStatusSuccess)
+			{
+				part = pitData->FindEntry(partitionIdentifier);
+
+				if (!part)
+				{
+					Interface::PrintError("No partition with identifier \"%s\" exists in the specified PIT.\n", it->argumentName);
+					return (false);
+
+				}
+			} else {
+				part = pitData->FindEntry(it->argumentName);
+
+				if (!part)
+				{
+					Interface::PrintError("Partition \"%s\" does not exist in the specified PIT.\n", it->argumentName);
+					return (false);
+
+				}
+			}
+
 			if (part->GetDeviceType() != PitEntry::kDeviceTypeMMC &&
 			    part->GetDeviceType() != PitEntry::kDeviceTypeUFS)
 				continue;
-- 
2.45.2

